File: data/record_manager.py
"""
RecordManager - Local RecordDB와 V-Archive 캐시를 머지하는 레이어
"""

from __future__ import annotations
from typing import Optional

from data.record_db import RecordDB
from data.steam_session import mask_steam_id
from data.varchive_client import VArchiveRecordClient


class RecordManager:
    def __init__(self, record_db: RecordDB, varchive_client: VArchiveRecordClient):
        self.rdb = record_db
        self.vclient = varchive_client
        self._varchive_cache: dict[tuple[int, str, str], tuple[float, bool]] = {}
        self._current_steam_id: Optional[str] = None
        self._data_revision: int = 0
        self._dirty_record_keys: set[tuple[int, str, str]] = set()
        self._full_dirty: bool = True

    def initialize(self) -> bool:
        success = self.rdb.initialize()
        self.refresh()
        return success

    def set_steam_id(self, steam_id: Optional[str]) -> tuple[bool, str, str]:
        changed, old_masked, new_masked = self.rdb.set_steam_id(steam_id)
        if changed:
            self.refresh()
        return changed, old_masked, new_masked

    def refresh(self):
        """현재 Steam ID에 해당하는 V-Archive 캐시를 메모리로 로드한다."""
        steam_id = self.rdb.get_steam_id()
        self._current_steam_id = steam_id
        self._varchive_cache = {}

        if steam_id == "__unknown__":
            return

        for button in [4, 5, 6, 8]:
            button_mode = f"{button}B"
            records = self.vclient.load_cached_records(steam_id, button)
            for rec in records:
                # rec: { "title": song_id, "pattern": "SC", "score": 99.99, ... }
                try:
                    song_id = int(rec.get("title", 0))
                    diff = rec.get("pattern")
                    rate = float(rec.get("score", 0.0))
                    is_max_combo = bool(rec.get("maxCombo", False))
                    if song_id is not None and diff:
                        self._varchive_cache[(song_id, button_mode, diff)] = (rate, is_max_combo)
                except (ValueError, TypeError):
                    continue
        
        print(f"[RecordManager] V-Archive 캐시 로드 완료: {len(self._varchive_cache)} 건 (steam_id={mask_steam_id(steam_id)})")
        self._data_revision += 1
        self._full_dirty = True
        self._dirty_record_keys.clear()

    def upsert(
        self, 
        song_id: int, 
        button_mode: str, 
        difficulty: str, 
        rate: float,
        is_max_combo: bool = False
    ) -> bool:
        """로컬 DB에 저장 (V-Archive 캐시는 읽기 전용)"""
        updated = self.rdb.upsert(song_id, button_mode, difficulty, rate, is_max_combo)
        if updated:
            self._data_revision += 1
            self._dirty_record_keys.add((song_id, button_mode, difficulty))
        return updated

    def delete(
        self, 
        song_id: int, 
        button_mode: str, 
        difficulty: str
    ) -> bool:
        """로컬 DB에서 기록 삭제"""
        deleted = self.rdb.delete(song_id, button_mode, difficulty)
        if deleted:
            self._data_revision += 1
            self._dirty_record_keys.add((song_id, button_mode, difficulty))
        return deleted

    def get(self, song_id: int, button_mode: str, difficulty: str) -> Optional[dict]:
        local_data = self.rdb.get(song_id, button_mode, difficulty)
        v_rate, is_max_combo = self._varchive_cache.get((song_id, button_mode, difficulty), (None, False))

        if local_data is None and v_rate is None:
            return None
            
        rate = v_rate if v_rate is not None else 0.0

        if local_data:
            rate = max(rate, local_data["rate"])
            is_max_combo = local_data["is_max_combo"] or is_max_combo
            
        return {
            "rate": rate,
            "is_max_combo": is_max_combo
        }

    def get_bulk(self, song_ids: list[int], button_mode: str, difficulty: str) -> dict[int, dict]:
        local_map = self.rdb.get_bulk(song_ids, button_mode, difficulty)
        result = local_map.copy()

        for sid in song_ids:
            v_rate, is_max_combo = self._varchive_cache.get((sid, button_mode, difficulty))
            if v_rate is not None:
                if sid not in result:
                    result[sid] = {
                        "rate": v_rate,
                        "is_max_combo": is_max_combo
                    }
                else:
                    entry = result[sid]
                    entry["rate"] = max(entry["rate"], v_rate)
                    entry["is_max_combo"] |= is_max_combo
        
        return result

    def get_rate_map(self, song_ids: list[int]) -> dict[tuple[int, str, str], dict]:
        local_map = self.rdb.get_rate_map(song_ids)
        result = local_map.copy()

        for (sid, mode, diff), (v_rate, is_max_combo) in self._varchive_cache.items():
            if sid in song_ids:
                key = (sid, mode, diff)
                if key not in result:
                    result[key] = {
                        "rate": v_rate,
                        "is_max_combo": is_max_combo
                    }
                else:
                    entry = result[key]
                    entry["rate"] = max(entry["rate"], v_rate)
                    entry["is_max_combo"] |= is_max_combo

        return result

    def stats(self) -> dict:
        s = self.rdb.stats()
        s["varchive_cached_count"] = len(self._varchive_cache)
        return s
    
    # 델리게이션 메서드들
    def get_steam_id(self) -> str:
        return self.rdb.get_steam_id()
    
    @property
    def is_ready(self) -> bool:
        return self.rdb.is_ready
    
    @property
    def masked_steam_id(self) -> str:
        return self.rdb.masked_steam_id

    @property
    def data_revision(self) -> int:
        return self._data_revision

    def consume_dirty_info(self) -> tuple[bool, set[tuple[int, str, str]]]:
        """(full_dirty, dirty_keys)를 반환하고 dirty 상태를 소모한다."""
        full_dirty = self._full_dirty
        dirty_keys = set(self._dirty_record_keys)
        self._full_dirty = False
        self._dirty_record_keys.clear()
        return full_dirty, dirty_keys

File: rust/overmax_data/src/service/record_manager.rs
use crate::store::record_db::RecordDB;
use overmax_core::{RecordKey, RecordValue};
use std::collections::{HashMap, HashSet};
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::sync::{Arc, Mutex};

pub trait RecordSource {
    fn is_ready(&self) -> bool;
    fn get_rate_map(&self, song_ids: &[i32]) -> HashMap<RecordKey, RecordValue>;
}

pub struct RecordManager {
    record_db: Arc<RecordDB>,
    varchive_cache: Mutex<HashMap<RecordKey, RecordValue>>,
    data_revision: AtomicU64,
    dirty_record_keys: Mutex<HashSet<RecordKey>>,
    full_dirty: AtomicBool,
}

impl RecordManager {
    pub fn new(record_db: Arc<RecordDB>) -> Self {
        Self {
            record_db,
            varchive_cache: Mutex::new(HashMap::new()),
            data_revision: AtomicU64::new(0),
            dirty_record_keys: Mutex::new(HashSet::new()),
            full_dirty: AtomicBool::new(true),
        }
    }

    pub fn refresh(&self) {
        let steam_id = self.record_db.get_steam_id();
        let cache = self
            .record_db
            .load_varchive_records(&steam_id)
            .unwrap_or_default();
        if let Ok(mut guard) = self.varchive_cache.lock() {
            *guard = cache;
        }
        self.full_dirty.store(true, Ordering::SeqCst);
        if let Ok(mut guard) = self.dirty_record_keys.lock() {
            guard.clear();
        }
        self.data_revision.fetch_add(1, Ordering::SeqCst);
    }

    pub fn set_steam_id(&self, steam_id: Option<&str>) -> (bool, String, String) {
        let result = self.record_db.set_steam_id(steam_id);
        if result.0 {
            self.refresh();
        }
        result
    }

    pub fn upsert(
        &self,
        song_id: i32,
        button_mode: &str,
        difficulty: &str,
        rate: f32,
        is_max_combo: bool,
        only_if_improved: bool,
    ) -> bool {
        if self.record_db.upsert(
            song_id,
            button_mode,
            difficulty,
            rate as f64,
            is_max_combo,
            only_if_improved,
        ) {
            if let Ok(mut guard) = self.dirty_record_keys.lock() {
                guard.insert((song_id, button_mode.to_string(), difficulty.to_string()));
            }
            self.data_revision.fetch_add(1, Ordering::SeqCst);
            return true;
        }
        false
    }

    pub fn delete(&self, song_id: i32, button_mode: &str, difficulty: &str) -> bool {
        if self.record_db.delete(song_id, button_mode, difficulty) {
            if let Ok(mut guard) = self.varchive_cache.lock() {
                guard.remove(&(song_id, button_mode.to_string(), difficulty.to_string()));
            }
            if let Ok(mut guard) = self.dirty_record_keys.lock() {
                guard.insert((song_id, button_mode.to_string(), difficulty.to_string()));
            }
            self.data_revision.fetch_add(1, Ordering::SeqCst);
            return true;
        }
        false
    }

    pub fn data_revision(&self) -> u64 {
        self.data_revision.load(Ordering::SeqCst)
    }

    pub fn consume_dirty_info(&self) -> (bool, HashSet<RecordKey>) {
        let full_dirty = self.full_dirty.swap(false, Ordering::SeqCst);
        let mut keys = HashSet::new();
        if let Ok(mut guard) = self.dirty_record_keys.lock() {
            std::mem::swap(&mut *guard, &mut keys);
        }
        (full_dirty, keys)
    }

    pub fn get_local_record(
        &self,
        song_id: i32,
        button_mode: &str,
        difficulty: &str,
    ) -> Option<RecordValue> {
        self.record_db.get(song_id, button_mode, difficulty)
    }

    pub fn get_varchive_cache_record(
        &self,
        song_id: i32,
        button_mode: &str,
        difficulty: &str,
    ) -> Option<RecordValue> {
        let guard = overmax_core::lock_or_recover(&self.varchive_cache);
        guard
            .get(&(song_id, button_mode.to_string(), difficulty.to_string()))
            .copied()
    }

    fn merge_varchive_cache(&self, result: &mut HashMap<RecordKey, RecordValue>, song_ids: &[i32]) {
        let cache = overmax_core::lock_or_recover(&self.varchive_cache);
        let song_ids_set: HashSet<i32> = song_ids.iter().copied().collect();
        for (key, &(v_rate, v_mc)) in cache.iter() {
            if !song_ids_set.contains(&key.0) {
                continue;
            }
            result
                .entry(key.clone())
                .and_modify(|entry| {
                    entry.0 = entry.0.max(v_rate);
                    entry.1 |= v_mc;
                })
                .or_insert((v_rate, v_mc));
        }
    }
}

impl RecordSource for RecordDB {
    fn is_ready(&self) -> bool {
        self.is_ready
    }

    fn get_rate_map(&self, song_ids: &[i32]) -> HashMap<RecordKey, RecordValue> {
        RecordDB::get_rate_map(self, song_ids)
    }
}

impl RecordSource for RecordManager {
    fn is_ready(&self) -> bool {
        self.record_db.is_ready
    }

    fn get_rate_map(&self, song_ids: &[i32]) -> HashMap<RecordKey, RecordValue> {
        let mut result = self.record_db.get_rate_map(song_ids);
        self.merge_varchive_cache(&mut result, song_ids);
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;
    use std::time::{SystemTime, UNIX_EPOCH};

    #[test]
    fn rate_map_merges_local_and_varchive_cache_by_best_rate() {
        let dir = test_dir("record-manager-merge");
        let db_path = dir.join("record.db");
        let cache_root = dir.join("varchive");
        let steam_id = "765611";
        std::fs::create_dir_all(cache_root.join(steam_id)).unwrap();

        let mut db = RecordDB::new(&db_path, Some(steam_id));
        assert!(db.initialize());
        assert!(db.upsert(42, "4B", "MX", 98.0, false, false));
        write_cache(&cache_root, steam_id);
        db.migrate_json_cache_to_db(&cache_root).unwrap();

        let db = Arc::new(db);
        let manager = RecordManager::new(db);
        manager.refresh();

        let map = manager.get_rate_map(&[42, 99]);

        assert_eq!(
            map.get(&(42, "4B".into(), "MX".into())),
            Some(&(99.5, true))
        );
        assert_eq!(
            map.get(&(99, "4B".into(), "SC".into())),
            Some(&(97.0, false))
        );

        let _ = std::fs::remove_dir_all(dir);
    }

    fn write_cache(cache_root: &Path, steam_id: &str) {
        let payload = json!({
            "records": [
                {"title": "42", "pattern": "MX", "score": 99.5, "maxCombo": true},
                {"title": "99", "pattern": "SC", "score": 97.0, "maxCombo": false}
            ]
        });
        std::fs::write(
            cache_root.join(steam_id).join("4.json"),
            payload.to_string(),
        )
        .unwrap();
    }

    fn test_dir(prefix: &str) -> PathBuf {
        let nanos = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_nanos();
        let dir = std::env::temp_dir().join(format!("{prefix}-{}-{nanos}", std::process::id()));
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(&dir).unwrap();
        dir
    }

    #[test]
    fn test_recommendation_caching_and_stats() {
        use crate::community::client::VArchiveDB;
        use crate::service::recommend::Recommender;

        let mut vdb = VArchiveDB::new();
        let song1_json = serde_json::json!({
            "name": "Song A",
            "title": "1",
            "composer": "Artist A",
            "dlcCode": "pack",
            "patterns": {
                "4B": {
                    "MX": {
                        "level": 15,
                        "floorName": "15.0"
                    }
                }
            }
        });
        let song2_json = serde_json::json!({
            "name": "Song B",
            "title": "2",
            "composer": "Artist B",
            "dlcCode": "pack",
            "patterns": {
                "4B": {
                    "MX": {
                        "level": 15,
                        "floorName": "15.0"
                    }
                }
            }
        });
        vdb.songs = vec![
            serde_json::from_value(song1_json).unwrap(),
            serde_json::from_value(song2_json).unwrap(),
        ];

        let dir = test_dir("recommend-stats-cache");
        let db_path = dir.join("record.db");
        let mut db = RecordDB::new(&db_path, None);
        assert!(db.initialize());

        assert!(db.upsert(1, "4B", "MX", 99.0, false, false));
        assert!(db.upsert(2, "4B", "MX", 97.0, false, false));

        let record_db = Arc::new(db);
        let record_manager = Arc::new(RecordManager::new(record_db));
        record_manager.refresh();

        let recommender = Recommender::new(Arc::new(vdb), record_manager);

        let result = recommender.recommend(1, "4B", "MX", 0.1, 10, true);

        assert_eq!(result.entries.len(), 1);
        assert_eq!(result.entries[0].song_id, 2);

        assert_eq!(result.total_count, 2);
        assert_eq!(result.has_record_count, 2);
        assert_eq!(result.avg_rate, 98.0);

        let _ = std::fs::remove_dir_all(dir);
    }

    #[test]
    fn test_record_manager_helpers() {
        let dir = test_dir("record-manager-helpers");
        let db_path = dir.join("record.db");
        let cache_root = dir.join("varchive");
        let steam_id = "765611";
        std::fs::create_dir_all(cache_root.join(steam_id)).unwrap();

        let mut db = RecordDB::new(&db_path, Some(steam_id));
        assert!(db.initialize());
        assert!(db.upsert(123, "5B", "SC", 99.80, true, false));
        write_cache(&cache_root, steam_id); // Writes MX/SC cache: MX=99.5, SC=97.0 for song 42/99
        db.migrate_json_cache_to_db(&cache_root).unwrap();

        let db = Arc::new(db);
        let manager = RecordManager::new(db);
        manager.refresh();

        // 1. Verify get_local_record
        assert_eq!(
            manager.get_local_record(123, "5B", "SC"),
            Some((99.80, true))
        );
        assert_eq!(manager.get_local_record(999, "4B", "NM"), None);

        // 2. Verify get_varchive_cache_record
        // Write cache has MX 99.5 for song 42
        assert_eq!(
            manager.get_varchive_cache_record(42, "4B", "MX"),
            Some((99.5, true))
        );
        assert_eq!(manager.get_varchive_cache_record(42, "4B", "NM"), None);

        let _ = std::fs::remove_dir_all(dir);
    }
}
